api/v1alpha3/types.go

```diff
@@ -294,6 +298,13 @@ func (s Subnets) FilterByZone(zone string) (res Subnets) {
 	return
 }
 
+// CNISpec defines configuration for CNI
+type CNISpec struct {
+	// IngressRules specify rules to apply to control plane and worker node security groups
+	// The source for the rule will be set to control plane and worker security group IDs
+	IngressRules IngressRules `json:"ingressRules,omitempty"`
```

Comment:
Do we want omitempty here?

Reply:
Do we want to require at least 1 rule? To avoid things like `cni: {}`

Reply:
Also just to note, the `IngressRules` has other fields as well which we mentioned they wouldn't be exposed, like `sourceSecurityGroupIds`

Reply:
0 vs 1+ rule - I'm 🤷‍♂️. Maybe someone would want to manage CNI rules with additionalSecurityGroups instead?

We should probably define our own scoped-down struct specifically for this (to exclude sourceSecurityGroupIds)?

Reply:
If they want to use a different mechanism, a user can set (or never set) `cni: nil`. If we allow for empty ingress rules, the spec would look something like `cni: { ingressRules: [] }` which seems odd

Reply:
Good point. Let's require at least 1

Reply:
Just checked the VPC CNI plugin. This would work through additional security groups and no rules in the CNI. I think we should therefore allow the zero length slice and respect it.

Reply:
Thanks for checking @randomvariable! How do additional security groups tie in here?

Reply:
By default, if there's SGs attached to the primary interface, the VPC CNI will consume them. Though, as there's a chance that they'll start conflicting with CAPI in managing the rules, you would probably create new security groups, use `AdditionalSecurityGroups` in the AWSMachineTemplate, and use the VPC CNI plugin's custom ENIConfig CRD to force it to use that SG. The VPC CNI plugin manages the rules by itself, so no extra rules are needed.

Reply:
Got it, thanks!

@gab-satchi could you please modify your PR to allow a non-nil CNI field with 0 rules (and make sure it does not default Calico in this scenario)?